File: java-pi-pantry/src/main/java/com/langrsoft/pi/pantry/ItemRetriever.java
package com.langrsoft.pi.pantry;

import com.langrsoft.util.HttpClient;

import java.io.IOException;

public class ItemRetriever {
    private static final String API_SERVER = "http://api.upcdatabase.org";
    public static final String UPC_API_KEY_PROPERTY_NAME = "upcdatabase_api_key";
    private String upcApiKey = System.getProperty(UPC_API_KEY_PROPERTY_NAME);
    private HttpClient httpClient;
    private ItemFactory itemFactory = new ItemFactory();

    public ItemRetriever(HttpClient client) {
        this.httpClient = client;
    }

    String url(String upcNumber) {
        return String.format("%s/json/%s/%s", API_SERVER, upcApiKey, upcNumber);
    }

    public Item retrieve(String upcNumber) {
        String json = httpClient.retrieveText(url(upcNumber));
        return itemFactory.create(json);
    }
}

File: csharp-pi-pantry/csharp-pi-pantry/Domain/ItemRetriever.cs
using Pipantry.Util;

namespace Pipantry.Domain
{
    public class ItemRetriever
    {
        private const string API_SERVER = "http://api.upcdatabase.org";
        public const string UPC_API_KEY_PROPERTY_NAME = "upcdatabase_api_key";
        private string upcApiKey =
            // TODO
            //System.getProperty(UPC_API_KEY_PROPERTY_NAME);
"a6f99f4c683845042ae27bbf3e36aeee";
        private Http httpClient;
        private ItemFactory itemFactory = new ItemFactory();

        public ItemRetriever(Http client)
        {
            httpClient = client;
        }

        public string Url(string upcNumber)
        {
            return $"{API_SERVER}/json/{upcApiKey}/{upcNumber}";
        }

        public Item Retrieve(string upcNumber)
        {
            string json = httpClient.RetrieveText(Url(upcNumber));
            return itemFactory.Create(json);
        }
    }
}
